File: apps/cli-rs/src/workspace/file_cache.rs
use crate::workspace::types::FileEntry;
use std::collections::HashMap;

#[derive(Default)]
struct WorkspaceFileCache {
    directories: HashMap<String, Vec<FileEntry>>,
}

#[derive(Default)]
pub struct WorkspaceFileCacheStore {
    workspaces: HashMap<String, WorkspaceFileCache>,
}

impl WorkspaceFileCacheStore {
    pub fn get_directory(&self, root: &str, rel_path: &str) -> Option<Vec<FileEntry>> {
        self.workspaces
            .get(root)
            .and_then(|cache| cache.directories.get(rel_path))
            .cloned()
    }

    pub fn store_directory(&mut self, root: &str, rel_path: String, entries: Vec<FileEntry>) {
        self.workspaces
            .entry(root.to_string())
            .or_default()
            .directories
            .insert(rel_path, entries);
    }

    pub fn clear_workspace(&mut self, root: &str) {
        self.workspaces.remove(root);
    }

    pub fn invalidate_paths(&mut self, root: &str, paths: &[String]) {
        let Some(workspace) = self.workspaces.get_mut(root) else {
            return;
        };

        for path in paths {
            let normalized = normalize_cache_path(path);
            workspace.directories.remove(&normalized);

            let prefix = if normalized.is_empty() {
                String::new()
            } else {
                format!("{normalized}/")
            };
            if normalized.is_empty() {
                workspace.directories.clear();
                continue;
            }

            workspace
                .directories
                .retain(|key, _| key != &normalized && !key.starts_with(&prefix));

            let parent = parent_cache_path(&normalized);
            workspace.directories.remove(&parent);
        }

        if workspace.directories.is_empty() {
            self.workspaces.remove(root);
        }
    }
}

pub fn normalize_cache_path(path: &str) -> String {
    path.trim().trim_matches('/').replace('\\', "/")
}

pub fn parent_cache_path(path: &str) -> String {
    let normalized = normalize_cache_path(path);
    normalized
        .rsplit_once('/')
        .map(|(parent, _)| parent.to_string())
        .unwrap_or_default()
}

File: apps/cli/internal/workspace/file_cache.go
package workspace

import "strings"

type workspaceFileCache struct {
	directories map[string][]FileEntry
}

type fileCacheStore struct {
	workspaces map[string]workspaceFileCache
}

func newFileCacheStore() fileCacheStore {
	return fileCacheStore{workspaces: make(map[string]workspaceFileCache)}
}

func (s *fileCacheStore) getDirectory(root string, path string) ([]FileEntry, bool) {
	workspaceCache, ok := s.workspaces[root]
	if !ok {
		return nil, false
	}
	entries, ok := workspaceCache.directories[normalizeCachePath(path)]
	if !ok {
		return nil, false
	}
	return cloneFileEntries(entries), true
}

func (s *fileCacheStore) storeDirectory(root string, path string, entries []FileEntry) {
	workspaceCache := s.workspaces[root]
	if workspaceCache.directories == nil {
		workspaceCache.directories = make(map[string][]FileEntry)
	}
	workspaceCache.directories[normalizeCachePath(path)] = cloneFileEntries(entries)
	s.workspaces[root] = workspaceCache
}

func (s *fileCacheStore) invalidatePaths(root string, paths []string) {
	workspaceCache, ok := s.workspaces[root]
	if !ok {
		return
	}
	for _, path := range paths {
		normalizedPath := normalizeCachePath(path)
		if normalizedPath == "" {
			delete(s.workspaces, root)
			return
		}
		prefix := normalizedPath + "/"
		for cachedPath := range workspaceCache.directories {
			if cachedPath == normalizedPath || strings.HasPrefix(cachedPath, prefix) {
				delete(workspaceCache.directories, cachedPath)
			}
		}
		delete(workspaceCache.directories, parentCachePath(normalizedPath))
	}
	if len(workspaceCache.directories) == 0 {
		delete(s.workspaces, root)
		return
	}
	s.workspaces[root] = workspaceCache
}

func normalizeCachePath(path string) string {
	return strings.Trim(strings.ReplaceAll(strings.TrimSpace(path), "\\", "/"), "/")
}

func parentCachePath(path string) string {
	normalizedPath := normalizeCachePath(path)
	separatorIndex := strings.LastIndex(normalizedPath, "/")
	if separatorIndex < 0 {
		return ""
	}
	return normalizedPath[:separatorIndex]
}

func cloneFileEntries(entries []FileEntry) []FileEntry {
	return append([]FileEntry(nil), entries...)
}
